lib/segment/src/index/field_index/geo_index.rs

```diff
@@ -1,572 +1,479 @@
-
-        let to_column = self.south_west[..precision].to_owned();
-
-        loop {
-            let mut current = from_row.clone();
-            loop {
-                seen.push(current.clone());
-
-                if seen.len() > max_regions {
-                    return None;
-                }
-
-                if current == to_row {
+impl PersistedGeoMapIndex {
+    pub fn match_cardinality(&self, values: &[GeoHash]) -> CardinalityEstimation {
+        let mut top_regions: HashSet<char> = HashSet::new();
+        values.iter().for_each(|geo_hash| {
+            let first_char = geo_hash.chars().next().unwrap();
+            top_regions.insert(first_char);
+        });
+
+        let mut all_possible_regions = HashSet::new();
+        values.iter().for_each(|geo_hash| {
+            // fallback to decreasing precision within the map if not found
+            decompose_geo_hash(geo_hash).for_each(|g| {
+                all_possible_regions.insert(g);
+            })
+        });
+
+        let mut values_count = 0;
+
+        // finds the appropriate precision to use to estimate each regions
+        top_regions.into_iter().for_each(|t| {
+            let sub_regions: Vec<_> = all_possible_regions
+                .iter()
+                .filter(|r| r.starts_with(t))
+                .collect();
+            let highest_precision = sub_regions
+                .iter()
+                .max_by(|a, b| a.len().cmp(&b.len()))
+                .unwrap()
+                .len();
+            for p in (0..=highest_precision).rev() {
+                let sum_region_count: usize = sub_regions
+                    .iter()
+                    .filter(|r| r.len() == p)
+                    .filter_map(|g| self.cardinality_map.get(&g.to_string()))
+                    .sum();
+                if sum_region_count != 0 {
+                    values_count += sum_region_count;
                     break;
                 }
-                current = sphere_neighbor(&current, Direction::E).unwrap();
             }
-            if from_row == to_column {
-                break;
-            }
-
-            from_row = sphere_neighbor(&from_row, Direction::S).unwrap();
-            to_row = sphere_neighbor(&to_row, Direction::S).unwrap();
+        });
+
+        let total_points_count = self.points_count;
+        let min = if total_points_count == self.payload_count {
+            values_count
+        } else {
+            // between 0 and 1 (there is at least one payload per point)
+            let payload_per_point = total_points_count as f64 / self.payload_count as f64;
+            values_count.saturating_sub((values_count as f64 * payload_per_point) as usize)
+        };
+        // don't overflow max number of points
+        let expected_count = total_points_count.min(values_count);
+        CardinalityEstimation {
+            primary_clauses: vec![],
+            min,
+            exp: expected_count,
+            max: expected_count,
```

Comment:
But `max` != `expected` in this case. For `max` we should guarantee, that it is impossible to match more than this amount of points with this query